domain/application/state/application.go

```diff
@@ -638,9 +645,48 @@ WHERE name = $unitDetails.name
 	return unit.LifeID, errors.Annotatef(err, "querying unit %q life", unitName)
 }
 
-// ApplicationScaleState looks up the scale state of the specified application, returning an error
+// SetUnitLife sets the life of the specified unit, returning an error
+// satisfying [applicationerrors.UnitNotFound] if the unit is not found.
+func (st *ApplicationState) SetUnitLife(ctx domain.AtomicContext, unitName string, l life.Life) error {
+	unit := coreUnit{Name: unitName, LifeID: l}
+	query := `
+SELECT uuid AS &coreUnit.uuid
+FROM unit
+WHERE name = $coreUnit.name
+`
+	stmt, err := st.Prepare(query, unit)
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	updateLifeQuery := `
+UPDATE unit
+SET life_id = $coreUnit.life_id
+WHERE name = $coreUnit.name
+-- we ensure the life can never go backwards.
+AND life_id < $coreUnit.life_id
+`
+
+	updateLifeStmt, err := st.Prepare(updateLifeQuery, unit)
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	err = domain.Run(ctx, func(ctx context.Context, tx *sqlair.TX) error {
+		err := tx.Query(ctx, stmt, unit).Get(&unit)
+		if errors.Is(err, sqlair.ErrNoRows) {
+			return fmt.Errorf("unit %q not found%w", unitName, errors.Hide(applicationerrors.UnitNotFound))
+		} else if err != nil {
+			return errors.Annotatef(err, "querying unit %q", unitName)
+		}
+		return tx.Query(ctx, updateLifeStmt, unit).Run()
```

Comment:
If the life was a backwards shift, wouldn't this change nothing, but return no error, meaning you could subsequently revoke leadership?

Reply:
Yes that is true, good pick up. I'll fix

Reply:
Actually, so for the case of EnsureDead, Dead is the "maximum" life value, so it can't go backwards.
0 = alive
1 = dying
2 = dead

The only way we get nothing being changed is if the unit was already dead and an attempt is made to set back to dying. But this does trigger revoke leadership - we only do that if setting life to Dead, so we never encounter the backwards noop issue as it pertains to falsely trigger revoke leadership.